### PR TITLE
[SYCL][ESIMD] Fix two opaque pointer issues

### PR DESCRIPTION
**Pull request by sarnex**
Issue 1) We need to change how we compute the offset for a GEP. This code is based on code from @jcranmer-intel

Issue 2) Missed case from https://github.com/intel/llvm/commit/637d3d5b674126f3945eee8e335cbfcb840628e9, just need to use the correct type.

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 1]
```diff
@@ -1237,10 +1237,19 @@ translateSpirvGlobalUses(LoadInst *LI, StringRef SpirvGlobalName,
     NewInst = llvm::Constant::getNullValue(LI->getType());
   } else if (isa<GetElementPtrConstantExpr>(LI->getPointerOperand())) {
     // Translate the load that has getelementptr as an operand
-    auto *GEPCE = cast<GetElementPtrConstantExpr>(LI->getPointerOperand());
-    uint64_t IndexValue =
-        cast<Constant>(GEPCE->getOperand(2))->getUniqueInteger().getZExtValue();
-    NewInst = generateSpirvGlobalGenX(LI, SpirvGlobalName, IndexValue);
+    auto *GEPCE = cast<GEPOperator>(LI->getPointerOperand());
```
Comment (turinevgeny) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:1240]: nit: should the name still have `CE` postfix, if it seems to be no longer a ConstantExpr?

**Review by turinevgeny — APPROVED** [review]

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp (hunk 1 above)
Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:1240]: good point will fix

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 2]
```diff
@@ -1237,10 +1237,19 @@ translateSpirvGlobalUses(LoadInst *LI, StringRef SpirvGlobalName,
     NewInst = llvm::Constant::getNullValue(LI->getType());
   } else if (isa<GetElementPtrConstantExpr>(LI->getPointerOperand())) {
     // Translate the load that has getelementptr as an operand
-    auto *GEPCE = cast<GetElementPtrConstantExpr>(LI->getPointerOperand());
-    uint64_t IndexValue =
-        cast<Constant>(GEPCE->getOperand(2))->getUniqueInteger().getZExtValue();
-    NewInst = generateSpirvGlobalGenX(LI, SpirvGlobalName, IndexValue);
+    auto *GEPOp = cast<GEPOperator>(LI->getPointerOperand());
```
Comment (jcranmer-intel) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp:1240]: You should be able to replace this with `} else if (auto *GEPOp = cast<GEPOperator>(LI->getPointerOperand())) {`

Comment (sarnex) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp]: fixed thx

**Comment by sarnex** [comment]
```
  Error: failed to start containers: 69212818171b9803b8135cfba3ceb2dbedfc1fc1b79a30e35c8f12ce830964a4
  Error: Docker start fail with exit code 1
```

cuda failure unrelated

@intel/llvm-gatekeepers can we merge pls? thx!

**Pull request merged**